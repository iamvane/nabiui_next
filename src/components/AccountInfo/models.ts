export enum Gender {
  male = 'male',
  female = 'female',
  undislcosed = 'undisclosed'
}

export enum VerificationChannel {
  Text = 'sms',
  Call = 'call'
}

export interface AccountInfoType {
  firstName: string;
  lastName: string;
  middleName?: string;
  gender?: Gender | string;
  phoneNumber?: string;
  location?: string;
  lat?: string;
  lng?: string;
  isPhoneVerified?: boolean;
}
