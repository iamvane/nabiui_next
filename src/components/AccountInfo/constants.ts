import { VerificationChannel } from './models';

export namespace AccountInfoComponent {
  export enum SectionTitles {
    Gender = 'Gender',
    Avatar = 'Profile Image',
    Location = 'Location'
  }

  export enum Labels {
    Gender = 'Gender',
    Female = 'Female',
    Male = 'Male',
    Birthday = 'Birthday'
  }

  export enum Placeholders {
    FirstName = 'First Name',
    LastName = 'Last Name',
    MiddleName = 'Middle Name',
  }

  export enum FieldKey {
    FirstName = 'firstName',
    LastName = 'lastName',
    MiddleName = 'middleName',
    Gender = 'gender',
    PhoneNumber = 'phoneNumber',
    Token = 'token',
    Location = 'location'
  }

  export const FieldNames = {
    [FieldKey.FirstName]: 'firstName',
    [FieldKey.LastName]: 'lastName',
    [FieldKey.MiddleName]: 'middleName',
    [FieldKey.Gender]: 'gender',
    [FieldKey.PhoneNumber]: 'phoneNumber',
    [FieldKey.Token]: 'token'
  };

  export interface Errors {
    [FieldKey.FirstName]?: string;
    [FieldKey.LastName]?: string;
    [FieldKey.MiddleName]?: string;
    [FieldKey.Gender]?: string;
    [FieldKey.PhoneNumber]?: string;
    [FieldKey.Token]?: string;
    [FieldKey.Location]?: string;

  }

  export const firstNameMinVal = 1;
  export const firstNameMaxVal = 50;

  export const middleNameMaxVal = 50;

  export const lastNameMinVal = 1;
  export const lastNameMaxVal = 50;

  export const errorMessages = {
    [FieldKey.FirstName]: {
      emptyFirstName: 'Enter your first name, up to {maxFirstNameValue} characters.'.replace(
        '{maxFirstNameValue}',
        firstNameMaxVal.toLocaleString()
      ),
      invalidFirstName: 'Enter a valid first name.'
    },
    [FieldKey.MiddleName]: {
      emptyMiddleName: 'Enter your middle name, up to {maxMiddleNameValue} characters.'.replace(
        '{maxMiddleNameValue}',
        middleNameMaxVal.toLocaleString()
      ),
      invalidMiddleName: 'Enter a valid middle name.'

    },
    [FieldKey.LastName]: {
      emptyLastName: 'Enter your last name, up to {maxLastNameValue} characters.'.replace(
        '{maxLastNameValue}',
        lastNameMaxVal.toLocaleString()
      ),
      invalidLastName: 'Enter a valid last name.'
    },
    [FieldKey.Token]: {
      invalidVerificationCode: 'Invalid verification code.',
      enterYourVerificationCode: 'Enter verification code.'
    },
    [FieldKey.Gender]: 'Select an option.',
    PhoneNumberEmpty: 'Enter a valid phone number.',
    PhoneNumberNotVerified: 'Verify phone number.',
    PhoneMissingCountryCode: 'Select country code.',
    [FieldKey.Location]: 'Enter a valid location.'
  };
}

export namespace PhoneValidationComponent {
  export const usPhoneNumberLenght = 10;
}

export namespace PhoneValidationFormComponent {
  export const sectionTitle = 'Verify Phone Number';

  export enum FieldNames {
    PhoneNumber = 'phoneNumber',
    PhoneToken = 'token',
    VefificationChannel = 'verificationChannel'
  }

  export enum Placeholders {
    PhoneNumber = 'Phone Number',
    PhoneToken = 'Verification Token'
  }

  export enum Buttons {
    Verify = 'Request Verification',
    Validate = 'Verify Phone',
    Cancel = 'Cancel'
  }

  export const resendCode = `Resend code or change phone number`;

  export const validationSucessful = 'Phone Verified';

  export const verificationMethod = {
    Text: {
      label: 'Text',
      value: VerificationChannel.Text
    },
    Call: {
      label: 'Call',
      value:  VerificationChannel.Call
    }
  };
}

export namespace LocationComponent {
  export const sectionTitle = 'Location';
  export const description = 'Please enter your home address.';
}

export namespace MiddleNameModalComponent {
  export const title = 'Missing Middle Name';
  export const message = 'Looks like you didnâ€™t provide a middle name. Please enter it below, if applicable.';

  export const noMiddleName = `I don't have one`;
  export const applyButton = 'Apply';
}
