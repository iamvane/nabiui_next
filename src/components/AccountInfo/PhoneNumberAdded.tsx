import * as React from 'react';

import {
  Grid,
  IconButton,
  Typography
} from '@material-ui/core';
// import Done from '@material-ui/icons/Done';
import dynamic from "next/dynamic";
const Done = dynamic(() => import('@material-ui/icons/Done'), {
  ssr: false,
});

import { PhoneValidationFormComponent } from './constants';

interface Props {
  phoneNumber: string | undefined;
  isEditing?: boolean;
}

const PhoneNumberAdded: React.StatelessComponent<Props> = props => {
  let value = props.phoneNumber + '';
  let formattedText = value.replace(/^(\d{3})(\d{3})(\d{4}).*/, '+1 ($1) $2-$3');

  return (
    <Grid item={true} xs={12} container={true}>
      <Grid xs={12} md={props.isEditing ? 5 : null} item={true}>
        <Typography>
          {formattedText}
        </Typography>
      </Grid>
      <Grid xs={12} md={props.isEditing ? 6 : null} item={true}>
        <Typography
          color="primary"
          className="nabi-margin-right-xsmall nabi-color-nabi nabi-display-inline"
        >
          {PhoneValidationFormComponent.validationSucessful}
        </Typography>
        <IconButton color="primary" disabled={true}>
          <Done />
        </IconButton>
      </Grid>
    </Grid>
  );
};

export default PhoneNumberAdded;
