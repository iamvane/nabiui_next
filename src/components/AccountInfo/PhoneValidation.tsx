import * as React from 'react';
import { connect } from 'react-redux';
import {
  Action,
  Dispatch
} from 'redux';

import { StoreState } from '../../redux/reducers/store';
import {
  requestToken,
  verifyToken
} from '../../redux/actions/UserActions';
import { UserActions } from '../../redux/actions/UserActionTypes';
import { requestAction } from '../../redux/actions/actions';
import SnackBar from '../common/SnackBar';
import PhoneValidationForm from './PhoneValidationForm';
import { AccountInfoComponent, PhoneValidationComponent } from './constants';
import { VerificationChannel } from './models';

interface DispatchProps {
  requestToken: (phoneNumber: string, channel: VerificationChannel) => void;
  verifyToken: (phoneNumber: string, token: string) => void;
  resetRequestTokenMessage: () => void;
}

interface StateProps {
  isRequestingToken: boolean;
  isVerifyingToken: boolean;
  errorRequestToken: string;
  errorVerifyToken: string;
  requestTokenMessage: string;
  verifyTokenMessage: string;
  phoneNumber: string;
  isPhoneVerified: boolean;
}

interface OwnProps {
  toggleChangePhoneNumber?: () => void;
  isEditing?: boolean;
  error?: string;
  customClass?: string;
  hideLabel?: boolean;
}

interface Props extends
  DispatchProps,
  StateProps,
  OwnProps {}

interface State {
  verificationChannel: string;
  isPhoneSet: boolean;
  token: string;
  errors: any;
  showSnackbar: boolean;
  snackbarMessage: string;
  phoneNumber: string;
  isPhoneVerified?: boolean;
}

export class PhoneValidation extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      verificationChannel: VerificationChannel.Text,
      isPhoneSet: false,
      token: '',
      errors: {},
      showSnackbar: false,
      snackbarMessage: '',
      phoneNumber: ''
    };
  }

  public componentDidMount() {
    if (this.props.phoneNumber) {
      this.setState({
        phoneNumber: this.props.phoneNumber
      })
    }
    if (this.props.isPhoneVerified) {
      this.setState({
        isPhoneVerified: this.props.isPhoneVerified
      })
    }
    if (this.props.requestTokenMessage) {
      this.setState({
        isPhoneSet: true
      })
    }
  }

  public handleNumberChange = (value: string): void => {
    this.setState({
      ...this.state,
      ...(Boolean(value) && {
        phoneNumber: value
      })
    });
  }

  public handleChange = (event: React.FormEvent<HTMLInputElement>): void => {
    const target = event.currentTarget;
    const value = target.value;
    const name = target.name;

    this.setState({
      ...this.state,
      [name]: value
    });
  }

  public componentDidUpdate(prevProps: Props): void {
    if ((prevProps.phoneNumber !== this.props.phoneNumber) && this.props.phoneNumber) {
      this.setState({
        ...this.state,
        phoneNumber: this.props.phoneNumber
      })
    }
    if ((prevProps.isPhoneVerified !== this.props.isPhoneVerified) && this.props.isPhoneVerified) {
      this.setState({
        ...this.state,
        isPhoneVerified: this.props.isPhoneVerified
      })
    }

    if ((prevProps.requestTokenMessage !== this.props.requestTokenMessage) && this.props.requestTokenMessage) {
      this.setState({
        ...this.state,
        showSnackbar: true,
        snackbarMessage: this.props.requestTokenMessage,
        isPhoneSet: true
      });
    }

    if ((prevProps.verifyTokenMessage !== this.props.verifyTokenMessage) && this.props.verifyTokenMessage) {
      this.setState({
        ...this.state,
        showSnackbar: true,
        snackbarMessage: this.props.verifyTokenMessage
      });
    }

    if ((prevProps.errorVerifyToken !== this.props.errorVerifyToken) && this.props.errorVerifyToken) {
      this.setState({
        ...this.state,
        errors: {...this.state.errors, [AccountInfoComponent.FieldKey.Token]: this.props.errorVerifyToken}
      });
    }
    if (!this.state.isPhoneSet && this.props.requestTokenMessage) {
      this.setState({
        ...this.state,
        isPhoneSet: true
      });
    }
  }

  public resendCode = () => {
    this.setState({
      isPhoneSet: false,
      token: ''
    });
    this.props.resetRequestTokenMessage();
  }

  public sendVerificationToken = async () => {
    const { errorMessages, FieldKey } = AccountInfoComponent;

    if ((document.querySelector("select[name='phoneNumberCountry']") as any).value === 'ZZ') {
      return this.setState({errors: {...this.state.errors, [FieldKey.PhoneNumber]: errorMessages.PhoneMissingCountryCode}});
    }

    if (!this.state.phoneNumber ||
      this.state.phoneNumber.length < PhoneValidationComponent.usPhoneNumberLenght) {
      return this.setState({errors: {...this.state.errors, [FieldKey.PhoneNumber]: errorMessages.PhoneNumberEmpty}});
    }

    await this.props.requestToken(
      this.state.phoneNumber,
      this.state.verificationChannel as VerificationChannel
    );

    if (this.props.errorRequestToken) {
      return this.setState({
        errors: {...this.state.errors, [FieldKey.PhoneNumber]: this.props.errorRequestToken}
      });
    }

    this.setState(prevState => ({
      isPhoneSet: !prevState.isPhoneSet,
      errors: {...this.state.errors, [FieldKey.Token]: ''},
    }));
  }

  public verifyToken = async () => {
    const { errorMessages, FieldKey } = AccountInfoComponent;
    const phoneNumber = this.state.phoneNumber || '';

    if (!this.state.token) {
      this.setState({
        errors: {...this.state.errors, [FieldKey.Token]: errorMessages.token.enterYourVerificationCode}
      });
    } else {
        await this.props.verifyToken(phoneNumber, this.state.token);
        if (this.props.errorVerifyToken) {
          return this.setState({
            errors: {...this.state.errors, [AccountInfoComponent.FieldKey.Token]: this.props.errorVerifyToken}
          });
        } else {
          this.setState({ isPhoneVerified: true });
        }
    }
  }

  public closeSnackbar = () => this.setState({ showSnackbar: false});

  public render(): JSX.Element {
    return (
      <div className={this.props.customClass ? this.props.customClass : ''}>
        <PhoneValidationForm
          phoneNumber={this.state.phoneNumber}
          isPhoneSet={this.state.isPhoneSet}
          handleChange={this.handleChange}
          handleNumberChange={this.handleNumberChange}
          token={this.state.token}
          sendVerificationToken={this.sendVerificationToken}
          verifyToken={this.verifyToken}
          verificationChannel={this.state.verificationChannel}
          toggleChangePhoneNumber={this.props.toggleChangePhoneNumber}
          resendCode={this.resendCode}
          isEditing={this.props.isEditing}
          errors={this.state.errors}
          isRequesting={this.props.isRequestingToken || this.props.isVerifyingToken}
          phoneError={this.props.error}
          isPhoneVerified={this.state.isPhoneVerified}
          hideLabel={this.props.hideLabel}
        />
        <SnackBar
          isOpen={this.state.showSnackbar}
          message={this.state.snackbarMessage}
          handleClose={this.closeSnackbar}
          variant="success"
        />
      </div>
    );
  }
}

const mapStateToProps = (state: StoreState, _ownProps: OwnProps): StateProps => {
  const {
    user: {
      phoneNumber,
      isPhoneVerified
    },
    actions: {
      requestToken: {
        isRequesting: isRequestingToken,
        error: errorRequestToken,
        message: requestTokenMessage
      },
      verifyToken: {
        isRequesting: isVerifyingToken,
        error: errorVerifyToken,
        message: verifyTokenMessage
      }
    },
  } = state.user;

  return {
    phoneNumber,
    isPhoneVerified,
    isRequestingToken,
    isVerifyingToken,
    errorRequestToken,
    errorVerifyToken,
    requestTokenMessage,
    verifyTokenMessage,
  };
};

const mapDispatchToProps = (
  dispatch: Dispatch<Action>
): DispatchProps => ({
  requestToken: (phoneNumber: string, channel: VerificationChannel) =>
    dispatch(requestToken(phoneNumber, channel)),
  verifyToken: (phoneNumber: string, token: string) =>
    dispatch(verifyToken(phoneNumber, token)),
  resetRequestTokenMessage: () => dispatch(requestAction(UserActions.RESET_REQUEST_TOKEN_MESSAGE))
});

export default connect(mapStateToProps, mapDispatchToProps)(PhoneValidation);
