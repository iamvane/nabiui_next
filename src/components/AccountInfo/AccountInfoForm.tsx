import * as React from 'react';

import 'react-datepicker/dist/react-datepicker.css';

import {
  FormLabel,
} from '@material-ui/core';

import {
  FormControl,
  FormControlLabel,
  FormHelperText,
  Radio,
  RadioGroup,
} from 'nabi_web_components';

import { LocationField } from "../Instructors/LocationField";
import SectionTitle from '../common/SectionTitle';
import { UserType } from '../../redux/models/UserModel';
import AvatarUploader from '../AvatarUploader/AvatarUploader';
import PhoneValidation from './PhoneValidation';
import { AccountInfoComponent } from './constants';
import {
  AccountInfoType,
  Gender,
} from './models';

interface Props {
  user: UserType;
  accountInfo: AccountInfoType;
  errors: AccountInfoComponent.Errors;
  redirectUrl: string;
  handleChange: (event: React.FormEvent<{}>) => void;
  getLatLng: (lat: string, lng: string) => void;
  handleLocationChange: (location: string) => void;
  location: string;
  showSections: string[];
  getLocationError: (error: string) => void;
}

const AccountInfoForm: React.StatelessComponent <Props> = props => {
    const {
      FieldNames,
      FieldKey
    } = AccountInfoComponent;

    const {
      gender,
    } = props.accountInfo;

    return(
      <div>
          <div className="nabi-margin-bottom-large">
            <div className="nabi-margin-bottom-small">
              <SectionTitle text={AccountInfoComponent.SectionTitles.Avatar} />
            </div>
            <AvatarUploader />
          </div>

        <div className={(props.showSections.includes('gender') || props.showSections.includes('showAll')) ? 'nabi-display-block' : 'nabi-display-none'}>
          <SectionTitle text={AccountInfoComponent.SectionTitles.Gender} />
          <div className="nabi-padding-left-small">
            <FormControl required={true} error={!!props.errors.gender}>
              <FormLabel className="nabi-margin-bottom-xsmall nabi-text-uppercase">
                {AccountInfoComponent.Labels.Gender}
              </FormLabel>

              <RadioGroup
                row={true}
                name={FieldNames[FieldKey.Gender]}
                onChange={props.handleChange}
                value={(gender && gender)}
              >
                <FormControlLabel
                  control={
                  <Radio
                    inputProps={{
                      name: Gender.female
                    }}
                  />}
                  label={AccountInfoComponent.Labels.Female}
                  value={Gender.female}
                />

                <FormControlLabel
                  control={
                    <Radio
                      inputProps={{
                        name: Gender.male
                      }}
                    />}
                  label={AccountInfoComponent.Labels.Male}
                  value={Gender.male}
                />
              </RadioGroup>

              <FormHelperText>{props.errors.gender}</FormHelperText>
            </FormControl>
          </div>
      </div>

      <div className={`nabi-margin-top-large ${(props.showSections.includes('phone') || props.showSections.includes('showAll')) ? 'nabi-display-block' : 'nabi-display-none'}`}>
        <PhoneValidation error={props.errors.phoneNumber} />
      </div>

      <div className={`nabi-margin-top-large ${(props.showSections.includes('location') || props.showSections.includes('showAll')) ? 'nabi-display-block' : 'nabi-display-none'}`}>
        <SectionTitle text={AccountInfoComponent.SectionTitles.Location} />
        <LocationField
          getLatLng={props.getLatLng}
          address={props.location}
          getLocation={props.handleLocationChange}
          getLocationError={props.getLocationError}
        />
      </div>
    </div>
  );
};

export default AccountInfoForm;
