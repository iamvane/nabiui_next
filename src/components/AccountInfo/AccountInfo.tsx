import * as React from 'react';
import Router from "next/router";
import { connect } from 'react-redux';
import {
  Action,
  Dispatch
} from 'redux';
import * as _ from "lodash";

import 'react-datepicker/dist/react-datepicker.css';

import { CircularProgress } from '@material-ui/core';
import dynamic from "next/dynamic";
const ArrowForward = dynamic(() => import('@material-ui/icons/ArrowForward'), {
  ssr: false,
});

import { StoreState } from '../../redux/reducers/store';
import {
  fetchUser,
  updateUser,
  changeAvatar
} from '../../redux/actions/UserActions';
import { UserType } from '../../redux/models/UserModel';
import { Routes } from '../common/constants/Routes';
import { CommonStepperButtons } from '../CommonStepper/constants';
import ConfirmExit from '../CommonStepper/ConfirmExit';
import { StepperButtons } from '../CommonStepper/StepperButtons';
import AccountInfoForm  from './AccountInfoForm';
import { AccountInfoComponent } from './constants';
import { AccountInfoType } from './models';
import { checkErrors } from "../../utils/checkErrors";

interface DispatchProps {
  fetchUser: () => void;
  updateUser: (user: Partial<AccountInfoType>) => void;
  changeAvatar: (id: string, avatar: string) => void;
}

interface OwnProps {
  nextPath: string;
}

interface StateProps {
  user: UserType;
  isRequestingFetch: boolean;
  isRequestingUpdate: boolean;
  errorUpdate: string;
  updateAvatarMessage: string;
}

interface Props extends
  DispatchProps,
  StateProps,
  OwnProps {
    hasImageUploader?: boolean;
    redirectUrl: string;
  }

  export const AccountInfo = (props: Props) => {
    const [accountInfo, setAccountInfo] = React.useState({} as AccountInfoType);
    const [errors, setErrors] = React.useState({} as AccountInfoComponent.Errors);
    const [performRedirect, setPerformRedirect] = React.useState(false);
    const [showSections, setShowSections] = React.useState(['avatar']);
    const [disableContinue, setDisableContinue] = React.useState(true);
    const [updateUser, setUpdateUser] = React.useState(false);
    const [displayExitModal, setDisplayExitModal] = React.useState(false);

    React.useEffect(() => {
      //get user
      const fetchData = async () => {
        await props.fetchUser();
      };
      fetchData();

      // set state
      if (props.user) {
        setAccountInfo({
          ...accountInfo,
          gender: accountInfo.gender || props.user.gender,
          location: props.user.location
        })
      }

      // if user data exists show all fields
      if (props.user.gender && props.user.avatar && props.user.location && props.user.isPhoneVerified) {
        setShowSections(['showAll'])
      }
      // updates active section when avatar is chanegd
      if (props.updateAvatarMessage || props.user.avatar || props.user.isPhoneVerified) {
        setDisableContinue(false);
      }

    },[props.updateAvatarMessage, props.user.gender, props.user.avatar, props.user.location, props.user.isPhoneVerified]);

    // if phone is already verified disable continue button
    React.useEffect(() => {
      if (props.user.isPhoneVerified && showSections.includes('phone') && !showSections.includes('showAll') && !showSections.includes('location')) {
        setDisableContinue(false);
      }
    },[showSections, props.user.isPhoneVerified]);

    // checks for updateUser state to make api call
    React.useEffect(() => {
      if (!props.errorUpdate && updateUser) {
        setPerformRedirect(true)
      }
    }, [updateUser, props.errorUpdate]);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const target = event.target;
    const value = target.value;
    const name = target.name;

    const { FieldNames } = AccountInfoComponent;

    if (name === FieldNames.gender) {
      setAccountInfo({
        ...accountInfo,
        gender: value
      })
      setDisableContinue(false);
    }
  };

  const validate = () => {
    const formErrors: AccountInfoComponent.Errors = {
    };

    const {
      errorMessages,
      FieldKey,
    } = AccountInfoComponent;
    // Validate gender
    if (!accountInfo.gender) {
      formErrors[FieldKey.Gender] = errorMessages[FieldKey.Gender];
    }

    // Validate phone number
    if (!props.user.isPhoneVerified) {
      formErrors[FieldKey.PhoneNumber] = errorMessages.PhoneNumberNotVerified;
    }

    // Validate location
    if (!accountInfo.location) {
      formErrors[FieldKey.Location] = errorMessages.location;
    }

    // Validate avatar
      if (!props.user.avatar) {
      formErrors[FieldKey.Avatar] = errorMessages.avatar;
    }
    return formErrors;
  }

  const handleLocationChange = (location: string) => {
    setAccountInfo({
      ...accountInfo,
      location
    })
    // clear location errors
    setErrors({
      ...errors,
      location: ''
    })
  }

  const getLatLng = (lat: string, lng: string) => {
    setAccountInfo({
      ...accountInfo,
      lat,
      lng
    })
    setDisableContinue(false);
  };

  const getLocationError = (error: string) => {
    setErrors({
      ...errors,
      location: error
    })
    setDisableContinue(true)
  }

  const currentState = {
    gender: accountInfo.gender,
    location: accountInfo.location
  }

  const storeState = {
    gender: props.user.gender,
    location: props.user.location
  }

  const handleExit = () => {
    if (showSections.includes('showAll') && _.isEqual(currentState, storeState)) {
      return Router.push(Routes.Dashboard);
    }
    setDisplayExitModal(true)
  }

  const handleNext = async() => {
    if (showSections.includes('avatar')) {
      setShowSections([
        'gender'
      ]),
      setDisableContinue(true);
    }

    if (showSections.includes('gender')) {
      setShowSections([
        ...showSections,
        'phone'
      ]),
      setDisableContinue(true);
    }

    if (showSections.includes('phone')) {
      setShowSections([
        ...showSections,
        'location'
      ]),
      setDisableContinue(true);
    }

    // if current state and store state are equal redirect
    if (showSections.includes('showAll') && _.isEqual(currentState, storeState)) {
      return setPerformRedirect(true);
    }

    if (showSections.includes('location') || showSections.includes('showAll')) {
      const formErrors = validate();
      const isError = checkErrors(Object.values(formErrors));

      if (isError) {
        return setErrors(formErrors);
      }

      // On no errors, call api and set update user state
      await props.updateUser({...accountInfo});
      setUpdateUser(true)
    }
  }

  return (
    props.isRequestingFetch || props.isRequestingUpdate ? (
      <div className="nabi-text-center">
        <CircularProgress />
      </div>
    ) : (
      <div>
        <AccountInfoForm
          user={props.user}
          accountInfo={accountInfo}
          errors={errors}
          redirectUrl={props.redirectUrl}
          handleChange={handleChange}
          getLatLng={getLatLng}
          getLocationError={getLocationError}
          handleLocationChange={handleLocationChange}
          location={accountInfo.location || ''}
          changeAvatar={props.changeAvatar}
          showSections={showSections}
        />
        {performRedirect && Router.push(props.redirectUrl)}
        <StepperButtons
          buttonText={CommonStepperButtons.Continue}
          handleNext={handleNext}
          icon={<ArrowForward />}
          isNextDisabled={disableContinue}
          errors={props.errorUpdate || Object.values(errors)}
          handleExit={handleExit}
        />
        <ConfirmExit
          isFormDialogOpen={displayExitModal}
          closeHandler={() => setDisplayExitModal(false)}
          handleProceed={() => Router.push(Routes.Dashboard)}
        />
      </div>
    )
  );
}

function mapStateToProps(state: StoreState, _ownProps: OwnProps): StateProps {
  const {
    user,
    actions: {
      fetchUser: {
        isRequesting: isRequestingFetch,
      },
      updateUser: {
        isRequesting: isRequestingUpdate,
        error: errorUpdate
      },
      uploadAvatar: {
        message: updateAvatarMessage
      }
    },
  } = state.user;

  return {
    user,
    isRequestingFetch,
    isRequestingUpdate,
    errorUpdate,
    updateAvatarMessage
  };
}

const mapDispatchToProps = (
  dispatch: Dispatch<Action>
): DispatchProps => ({
  fetchUser: () => dispatch(fetchUser()),
  updateUser: (user: AccountInfoType) => dispatch(updateUser(user)),
  changeAvatar: (id: string, avatar: string) => dispatch(changeAvatar(id, avatar)),
});

export default connect(mapStateToProps, mapDispatchToProps)(AccountInfo);
