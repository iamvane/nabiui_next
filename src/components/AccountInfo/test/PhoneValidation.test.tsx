import * as React from 'react';

import { shallow } from 'enzyme';
import { PhoneValidation } from 'components/AccountInfo/PhoneValidation';
import { UserType } from 'redux/models/UserModel';
import { accountInfo } from 'components/AccountInfo/test/mockData';

describe('PhoneValidation', () => {
  let wrapper: any;
  const mockFunction = jest.fn();

  const user: UserType = {
    ...accountInfo,
    id: 'fooId',
    email: 'email@company.com',
    password: 'secretPass123',
    role: 'Instructor'
  };

  beforeEach(() => {
    wrapper = shallow(
      <PhoneValidation
        user={user}
        isRequestingToken={false}
        isVerifyingToken={false}
        errorRequestToken=""
        errorVerifyToken=""
        requestTokenMessage=""
        verifyTokenMessage=""
        fetchUser={mockFunction}
        requestToken={mockFunction}
        verifyToken={mockFunction}
        resetRequestTokenMessage={mockFunction}
      />
    );
  });

  it('Renders the component', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
