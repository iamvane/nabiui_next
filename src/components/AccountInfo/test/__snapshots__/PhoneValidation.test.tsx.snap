// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhoneValidation Renders the component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PhoneValidation
    errorRequestToken=""
    errorVerifyToken=""
    fetchUser={[MockFunction]}
    isRequestingToken={false}
    isVerifyingToken={false}
    requestToken={[MockFunction]}
    requestTokenMessage=""
    resetRequestTokenMessage={[MockFunction]}
    user={
      Object {
        "email": "email@company.com",
        "firstName": "",
        "gender": "",
        "id": "fooId",
        "lastName": "",
        "lat": "",
        "lng": "",
        "location": "",
        "middleName": "",
        "password": "secretPass123",
        "phone": Object {
          "isVerified": false,
          "phoneNumber": "",
        },
        "phoneNumber": "",
        "referralToken": "",
        "role": "Instructor",
      }
    }
    verifyToken={[MockFunction]}
    verifyTokenMessage=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <PhoneValidationForm
          errors={Object {}}
          handleChange={[Function]}
          isEditing={undefined}
          isPhoneSet={false}
          isPhoneVerified={true}
          isRequesting={false}
          phoneError={undefined}
          phoneNumber=""
          resendCode={[Function]}
          sendVerificationToken={[Function]}
          toggleChangePhoneNumber={undefined}
          token=""
          user={
            Object {
              "email": "email@company.com",
              "firstName": "",
              "gender": "",
              "id": "fooId",
              "lastName": "",
              "lat": "",
              "lng": "",
              "location": "",
              "middleName": "",
              "password": "secretPass123",
              "phone": Object {
                "isVerified": false,
                "phoneNumber": "",
              },
              "phoneNumber": "",
              "referralToken": "",
              "role": "Instructor",
            }
          }
          verificationChannel="sms"
          verifyToken={[Function]}
        />,
        <SnackBar
          handleClose={[Function]}
          isOpen={false}
          message=""
          variant="success"
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "errors": Object {},
          "handleChange": [Function],
          "isEditing": undefined,
          "isPhoneSet": false,
          "isPhoneVerified": true,
          "isRequesting": false,
          "phoneError": undefined,
          "phoneNumber": "",
          "resendCode": [Function],
          "sendVerificationToken": [Function],
          "toggleChangePhoneNumber": undefined,
          "token": "",
          "user": Object {
            "email": "email@company.com",
            "firstName": "",
            "gender": "",
            "id": "fooId",
            "lastName": "",
            "lat": "",
            "lng": "",
            "location": "",
            "middleName": "",
            "password": "secretPass123",
            "phone": Object {
              "isVerified": false,
              "phoneNumber": "",
            },
            "phoneNumber": "",
            "referralToken": "",
            "role": "Instructor",
          },
          "verificationChannel": "sms",
          "verifyToken": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "handleClose": [Function],
          "isOpen": false,
          "message": "",
          "variant": "success",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <PhoneValidationForm
            errors={Object {}}
            handleChange={[Function]}
            isEditing={undefined}
            isPhoneSet={false}
            isPhoneVerified={true}
            isRequesting={false}
            phoneError={undefined}
            phoneNumber=""
            resendCode={[Function]}
            sendVerificationToken={[Function]}
            toggleChangePhoneNumber={undefined}
            token=""
            user={
              Object {
                "email": "email@company.com",
                "firstName": "",
                "gender": "",
                "id": "fooId",
                "lastName": "",
                "lat": "",
                "lng": "",
                "location": "",
                "middleName": "",
                "password": "secretPass123",
                "phone": Object {
                  "isVerified": false,
                  "phoneNumber": "",
                },
                "phoneNumber": "",
                "referralToken": "",
                "role": "Instructor",
              }
            }
            verificationChannel="sms"
            verifyToken={[Function]}
          />,
          <SnackBar
            handleClose={[Function]}
            isOpen={false}
            message=""
            variant="success"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "errors": Object {},
            "handleChange": [Function],
            "isEditing": undefined,
            "isPhoneSet": false,
            "isPhoneVerified": true,
            "isRequesting": false,
            "phoneError": undefined,
            "phoneNumber": "",
            "resendCode": [Function],
            "sendVerificationToken": [Function],
            "toggleChangePhoneNumber": undefined,
            "token": "",
            "user": Object {
              "email": "email@company.com",
              "firstName": "",
              "gender": "",
              "id": "fooId",
              "lastName": "",
              "lat": "",
              "lng": "",
              "location": "",
              "middleName": "",
              "password": "secretPass123",
              "phone": Object {
                "isVerified": false,
                "phoneNumber": "",
              },
              "phoneNumber": "",
              "referralToken": "",
              "role": "Instructor",
            },
            "verificationChannel": "sms",
            "verifyToken": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "handleClose": [Function],
            "isOpen": false,
            "message": "",
            "variant": "success",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
